File: examples_cpp/tests/controllers/test_pid_trajectory_geometric.cpp
// Copyright 2025 Universidad Politécnica de Madrid
// SPDX-License-Identifier: BSD-3-Clause

#include <gtest/gtest.h>

#include <cmath>
#include <memory>

#include "controllers/pid_trajectory_geometric_controller.hpp"
#include "framework/controller_base.hpp"
#include "test_helpers.hpp"

using mpc_examples::adapters::PidTrajectoryGeometricController;
using mpc_examples::framework::hasField;
using mpc_examples::framework::ReferenceField;
using mpc_examples::testing::horizonAtPosition;
using mpc_examples::testing::hoverStateAt;
using mpc_examples::testing::loadTestSimConfig;
using mpc_examples::testing::repoPath;

class PidTrajectoryGeometricControllerTest : public ::testing::Test {
protected:
  mpc_examples::ExampleConfig sim_cfg_ = loadTestSimConfig();
  mav_model::State state_              = hoverStateAt({0.0, 0.0, 10.0});
  std::unique_ptr<PidTrajectoryGeometricController> ctrl_;

  void SetUp() override {
    auto cfg = PidTrajectoryGeometricController::loadConfigFromYaml(
        repoPath("configs/controllers/config_pid_trajectory.yaml"));
    ctrl_ = std::make_unique<PidTrajectoryGeometricController>(cfg);
    ctrl_->initialize(state_, sim_cfg_);
  }
};

TEST_F(PidTrajectoryGeometricControllerTest, HorizonShapeIsSingleStep) {
  EXPECT_EQ(ctrl_->referenceHorizonSize(), 1);
  EXPECT_GT(ctrl_->controlPeriod(), 0.0);
}

TEST_F(PidTrajectoryGeometricControllerTest, RequiresPositionAndVelocityReference) {
  const auto mask = ctrl_->requiredReferenceFields();
  EXPECT_TRUE(hasField(mask, ReferenceField::kPosition));
  EXPECT_TRUE(hasField(mask, ReferenceField::kVelocity));
}

TEST_F(PidTrajectoryGeometricControllerTest, ProducesFiniteCommandOverBriefSimulation) {
  const auto refs = horizonAtPosition({5.0, 0.0, 10.0}, ctrl_->referenceHorizonSize());
  for (int k = 0; k < 50; ++k) {
    const auto cmd = ctrl_->computeCommand(state_, refs);
    ASSERT_TRUE(std::isfinite(cmd.thrust_n)) << "step " << k;
    ASSERT_TRUE(cmd.angular_rate.allFinite()) << "step " << k;
    EXPECT_GT(cmd.thrust_n, 0.0) << "step " << k;
  }
  EXPECT_GT(ctrl_->lastSolveTimeMicros(), 0.0);
}

// Replicate the worst-case waypoint-transition seen in the mav_traj_gen run:
// drone flying diagonally at ~3 m/s, new segment starts with ref_vel = 0.
// Velocity error ≈ [0, 2.8, 2.8] → Kd * error ≈ [0, 16.8, 16.8] m/s².
TEST_F(PidTrajectoryGeometricControllerTest,
       FiniteCommandAtWaypointTransitionWithHighVelocityError) {
  mav_model::State moving = hoverStateAt({0.0, 10.0, 20.0});
  moving.setLinearVelocityVector(Eigen::Vector3d(0.0, -2.8, -2.8));

  mpc_examples::framework::ReferenceSample s;
  s.position     = Eigen::Vector3d(0.0, 10.0, 20.0);
  s.velocity     = Eigen::Vector3d::Zero();
  s.acceleration = Eigen::Vector3d::Zero();
  s.yaw          = 0.0;
  const std::vector<mpc_examples::framework::ReferenceSample> refs(1, s);

  for (int k = 0; k < 50; ++k) {
    const auto cmd = ctrl_->computeCommand(moving, refs);
    ASSERT_TRUE(std::isfinite(cmd.thrust_n)) << "step " << k;
    ASSERT_TRUE(cmd.angular_rate.allFinite()) << "step " << k;
    EXPECT_GT(cmd.thrust_n, 0.0) << "step " << k;
  }
}

File: examples_py/tests/controllers/test_pid_trajectory_geometric.py
#!/usr/bin/env python3

# Copyright 2025 Universidad Politécnica de Madrid
# SPDX-License-Identifier: BSD-3-Clause

"""Unit test for :class:`examples_py.controllers.PidTrajectoryGeometricController`."""

import math
import unittest

import numpy as np

from examples_py.controllers import PidTrajectoryGeometricController
from examples_py.framework import ReferenceField, ReferenceSample, has_field
from _helpers import (
    hover_state_at,
    load_test_sim_config,
    repo_path,
)


class PidTrajectoryGeometricControllerTest(unittest.TestCase):
    def setUp(self) -> None:
        self.sim_cfg = load_test_sim_config()
        self.state = hover_state_at([0.0, 0.0, 10.0])
        cfg = PidTrajectoryGeometricController.load_config_from_yaml(
            str(repo_path('configs/controllers/config_pid_trajectory.yaml')))
        self.ctrl = PidTrajectoryGeometricController(cfg)
        self.ctrl.initialize(self.state, self.sim_cfg)

    def test_horizon_is_single_step(self) -> None:
        self.assertEqual(self.ctrl.reference_horizon_size(), 1)
        self.assertGreater(self.ctrl.control_period(), 0.0)

    def test_requires_position_and_velocity_reference(self) -> None:
        mask = self.ctrl.required_reference_fields()
        self.assertTrue(has_field(mask, ReferenceField.POSITION))
        self.assertTrue(has_field(mask, ReferenceField.VELOCITY))

    def test_produces_finite_command_over_brief_simulation(self) -> None:
        s = ReferenceSample()
        s.position = np.array([5.0, 0.0, 10.0])
        s.velocity = np.zeros(3)
        s.acceleration = np.zeros(3)
        s.yaw = 0.0
        refs = [s]
        for k in range(50):
            cmd = self.ctrl.compute_command(self.state, refs)
            self.assertTrue(math.isfinite(cmd.thrust_n), f'step {k}: thrust non-finite')
            self.assertTrue(np.all(np.isfinite(cmd.angular_rate)), f'step {k}: rates non-finite')
            self.assertGreater(cmd.thrust_n, 0.0, f'step {k}: thrust not positive')

    def test_finite_command_at_waypoint_transition_high_velocity_error(self) -> None:
        """Replicate the worst-case seen in mav_traj_gen run.

        Drone flying diagonally at ~3 m/s; new segment starts with ref_vel = 0.
        Velocity error ≈ [0, 2.8, 2.8] → Kd * error ≈ [0, 16.8, 16.8] m/s².
        """
        from mavpy.model import State
        moving = State()
        moving.position = [0.0, 10.0, 20.0]
        moving.orientation = [1.0, 0.0, 0.0, 0.0]
        moving.linear_velocity = [0.0, -2.8, -2.8]

        s = ReferenceSample()
        s.position = np.array([0.0, 10.0, 20.0])
        s.velocity = np.zeros(3)
        s.acceleration = np.zeros(3)
        s.yaw = 0.0
        refs = [s]

        for k in range(50):
            cmd = self.ctrl.compute_command(moving, refs)
            self.assertTrue(math.isfinite(cmd.thrust_n),
                            f'step {k}: thrust non-finite (acc_des ~[0,16.8,16.8] m/s²)')
            self.assertTrue(np.all(np.isfinite(cmd.angular_rate)),
                            f'step {k}: rates non-finite')
            self.assertGreater(cmd.thrust_n, 0.0, f'step {k}: thrust not positive')


if __name__ == '__main__':  # pragma: no cover
    unittest.main()
